### PR TITLE
Patrick Duvall

### PR DESCRIPTION
**Pull request by Patrick-Duvall**
IT4

### CODE REVIEW

File: lib/user.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  def learn(joke)
+    @jokes << joke
+  end
+
+  def tell(user, joke)
+
+    user.learn(joke)
+    @jokes_told << joke
+  end
+#could act wierd if multiple jokes have same id, I might use sympols for jokes, also, comedians are more likely to mentally label their jokes as say :dog_joke
```
Comment (ameseee) [inline lib/user.rb]: Good point!

Comment (ameseee) [inline lib/user.rb]: Good point!

File: lib/user.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+  end
+
+  def learn(joke)
+    @jokes << joke
+  end
+
+  def tell(user, joke)
+
+    user.learn(joke)
+    @jokes_told << joke
```
Comment (ameseee) [inline lib/user.rb]: I think we only want to add this joke to the user object that is being passed in; by calling `.learn` on it, we are the shoveling into its array of jokes. However, we don't necessarily want to put this joke into the "teller"'s array, so would want to remove line 16.

File: lib/user.rb [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    @jokes = []
+    @jokes_told = []
+  end
+
+  def learn(joke)
+    @jokes << joke
+  end
+
+  def tell(user, joke)
+
```
Comment (ameseee) [inline lib/user.rb]: Make sure to remove this whitespace

File: lib/open_mic.rb [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+  def welcome(user)
+    @performers << user
+  end
+# I feel I'm missing something obvious. This does add some functionality in that it gives users informaion on what jokes they've told.
+  def repeated_jokes?
+    specific_jokes = @performers.map(&:jokes_told).flatten.uniq
+    all_jokes = @performers.map(&:jokes_told).flatten
+    specific_jokes != all_jokes
+  end
+
```
Comment (ameseee) [inline lib/open_mic.rb]: make sure to only have one line of whitespace here😊

File: lib/joke.rb [hunk 5]
```diff
@@ -0,0 +1,10 @@
+class Joke
+  attr_reader :id, :setup, :punchline
+  def initialize(id, setup, punchline)
+    @id = id
+    @setup = setup
+    @punchline = punchline
+  end
+
+
```
Comment (ameseee) [inline lib/joke.rb]: Make sure to remove this whitespace!

File: lib/open_mic.rb [hunk 6]
```diff
@@ -0,0 +1,22 @@
+class OpenMic
+  attr_reader :location, :date, :performers
+  def initialize(attributes_hash)
+    @location = attributes_hash[:location]
+    @date = attributes_hash[:date]
+    @performers = []
+    # @repeated_jokes = false
```
Comment (ameseee) [inline lib/open_mic.rb]: not using so I'd delete this!

File: lib/open_mic.rb [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+  end
+
+  def welcome(user)
+    @performers << user
+  end
+# I feel I'm missing something obvious. This does add some functionality in that it gives users informaion on what jokes they've told.
```
Comment (ameseee) [inline lib/open_mic.rb]: I think you did great! Lots of solutions to writing this method; I think adding some state to the user is a fine way to do it.

File: test/joke_test.rb [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+require "minitest/autorun"
+
+class JokeTest<Minitest::Test
+#I was reading an article that argued that in an object test tests that object so calling it my name as test_joke_has_attributes is redundant. It continued this logic could be extended to test_it_has_attributes. I found it compelling.
+  def test_has_attributes
+    joke = Joke.new(1, "Why did the strawberry cross the road?", "Because his mother was in a jam.")
+    assert_equal 1, joke.id
+    assert_equal "Why did the strawberry cross the road?", joke.setup
+    assert_equal "Because his mother was in a jam.", joke.punchline
+
```
Comment (ameseee) [inline test/joke_test.rb]: beep delete me!

File: test/open_mic_test.rb [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    assert_equal "Comedy Works", @open_mic.location
+    assert_equal "11-20-18", @open_mic.date
+  end
+
+  def test_can_add_performers
+    assert_equal [], @open_mic.performers
+    @open_mic.welcome(@sal)
+    assert_equal [@sal], @open_mic.performers
+    @open_mic.welcome(@ali)
+    assert_equal [@sal, @ali], @open_mic.performers
```
Comment (ameseee) [inline test/open_mic_test.rb]: ✅

File: test/user_test.rb [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+    assert_equal [], @sal.jokes
+    @sal.learn(@joke_1)
+    assert_equal [@joke_1], @sal.jokes
+    @sal.learn(@joke_2)
+    assert_equal [@joke_1, @joke_2], @sal.jokes
+
+  end
+
+  def test_can_tell_jokes
+      @sal.learn(@joke_1)
```
Comment (ameseee) [inline test/user_test.rb]: These look like they are tabbed in twice from the def, only want one tab!

File: test/user_test.rb [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+      assert_equal [@joke_1], @ali.jokes
+      @sal.learn(@joke_2)
+      @sal.tell(@ali, @joke_2)
+      assert_equal [@joke_1, @joke_2], @ali.jokes
+  end
+
+  def test_knows_jokes_told
+    @sal.learn(@joke_1)
+    @sal.tell(@ali, @joke_1)
+    pry
```
Comment (ameseee) [inline test/user_test.rb]: make sure to remove pry

**Comment by ameseee** [comment]
Looks great @Patrick-Duvall! A few comments for your review - please let me know if you'd like to follow up on any of them.

**Comment by Patrick-Duvall** [comment]
Everything made sense, thanks for the feedback.

**Pull request closed without merging**